File: src/main/resources/static/js/webauthn.js
const webauthn = {
    async register(username, displayName, role) {
        const optsRes = await fetch('/api/auth/registration-options', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({ username, displayName })
        });
        const options = await optsRes.json();
        const credential = await navigator.credentials.create({
            publicKey: this.parseCreationOptions(options)
        });
        const regRes = await fetch('/api/auth/register', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({
                username, displayName, role,
                registrationResponse: JSON.stringify(credential.toJSON?.() ?? this.toJSON(credential))
            }),
            credentials: 'include'
        });
        const result = await regRes.json();
        return result.success;
    },
    async authenticate() {
        const optsRes = await fetch('/api/auth/assertion-options', { credentials: 'include' });
        const options = await optsRes.json();
        const credential = await navigator.credentials.get({
            publicKey: this.parseRequestOptions(options)
        });
        const authRes = await fetch('/api/auth/authenticate', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(credential.toJSON?.() ?? this.toJSON(credential)),
            credentials: 'include'
        });
        const result = await authRes.json();
        return result.success;
    },
    parseCreationOptions(opts) {
        if (window.PublicKeyCredential?.parseCreationOptionsFromJSON) {
            return PublicKeyCredential.parseCreationOptionsFromJSON(opts);
        }
        return {
            challenge: this.base64ToArray(opts.challenge),
            rp: opts.rp,
            user: { ...opts.user, id: this.base64ToArray(opts.user.id) },
            pubKeyCredParams: opts.pubKeyCredParams || [{ type: 'public-key', alg: -7 }, { type: 'public-key', alg: -257 }]
        };
    },
    parseRequestOptions(opts) {
        if (window.PublicKeyCredential?.parseRequestOptionsFromJSON) {
            return PublicKeyCredential.parseRequestOptionsFromJSON(opts);
        }
        return {
            challenge: this.base64ToArray(opts.challenge),
            rpId: opts.rpId || 'localhost'
        };
    },
    base64ToArray(s) {
        const bin = atob(s.replace(/-/g, '+').replace(/_/g, '/'));
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; i++) arr[i] = bin.charCodeAt(i);
        return arr;
    },
    arrayToBase64(arr) {
        let bin = '';
        for (let i = 0; i < arr.length; i++) bin += String.fromCharCode(arr[i]);
        return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
    },
    toJSON(cred) {
        const rawId = this.arrayToBase64(new Uint8Array(cred.rawId));
        const response = cred.response;
        return {
            id: cred.id,
            rawId,
            type: cred.type,
            response: {
                clientDataJSON: this.arrayToBase64(new Uint8Array(response.clientDataJSON)),
                authenticatorData: response.authenticatorData ? this.arrayToBase64(new Uint8Array(response.authenticatorData)) : undefined,
                signature: this.arrayToBase64(new Uint8Array(response.signature)),
                userHandle: response.userHandle ? this.arrayToBase64(new Uint8Array(response.userHandle)) : undefined
            }
        };
    }
};

File: frontend/src/jsMain/kotlin/com/adoptu/frontend/WebAuthn.kt
package com.adoptu.frontend

import kotlinx.browser.localStorage
import kotlinx.browser.window
import org.w3c.dom.HTMLElement
import kotlin.js.Promise
import kotlin.js.json

external fun encodeURIComponent(str: String): String

@JsExport
@JsName("webauthn")
object WebAuthnModule {
    private val navigator: dynamic
        get() = window.asDynamic().navigator

    fun register(email: String, displayName: String): Promise<dynamic> {
        console.log("Starting passkey registration for: $email")
        return getRegistrationOptions(email, displayName)
            .then<dynamic> { options ->
                console.log("Got registration options, creating credential")
                val publicKey = parseCreationOptions(options)
                try {
                    navigator.credentials.create(publicKey)
                } catch (e: dynamic) {
                    console.log("Credential creation error: $e")
                    throw e
                }
            }
            .then<dynamic> { credential ->
                if (credential == null) {
                    console.log("Passkey registration was cancelled by user")
                    throw js("new Error('Passkey registration was cancelled. Please try again or use password registration.')")
                }
                console.log("Got credential, encoding: " + credential)
                val idBase64 = arrayBufferToBase64(credential.id)
                val rawIdBase64 = arrayBufferToBase64(credential.rawId)
                val clientDataJSON = credential.response.clientDataJSON
                val clientBytes = js("new Uint8Array(clientDataJSON)")
                var clientBinary = ""
                for (i in 0 until clientBytes.length) {
                    clientBinary += js("String.fromCharCode(clientBytes[i])")
                }
                val clientBase64 = toBase64Url(window.btoa(clientBinary))
                val attestationObject = credential.response.attestationObject
                val attBytes = js("new Uint8Array(attestationObject)")
                var attBinary = ""
                for (i in 0 until attBytes.length) {
                    attBinary += js("String.fromCharCode(attBytes[i])")
                }
                val attBase64 = toBase64Url(window.btoa(attBinary))
                val jsonStr = """{"id":"$idBase64","type":"public-key","rawId":"$rawIdBase64","response":{"clientDataJSON":"$clientBase64","attestationObject":"$attBase64"}}"""
                console.log("Encoded json: " + jsonStr)
                val body = "email=${encodeURIComponent(email)}&displayName=${encodeURIComponent(displayName)}&registrationResponse=${encodeURIComponent(jsonStr)}"
                window.asDynamic().fetch("/api/auth/register", js("({method: 'POST', headers: {'Content-Type': 'application/x-www-form-urlencoded'}, body: body})")).then { res ->
                    console.log("Server responded: " + res.status)
                    res.unsafeCast<dynamic>().json().then<dynamic> { json -> json }
                }
            }
    }

    fun authenticate(): Promise<dynamic> {
        return getAuthenticationOptions()
            .then<dynamic> { options ->
                val publicKey = parseAssertionOptions(options)
                navigator.credentials.get(json("publicKey" to publicKey))
            }
            .then<dynamic> { credential ->
                if (credential == null) {
                    console.log("Passkey authentication was cancelled by user")
                    throw js("new Error('Passkey login was cancelled. Please try again or use password login.')")
                }
                val idBase64 = arrayBufferToBase64(credential.id)
                val rawIdBase64 = arrayBufferToBase64(credential.rawId)
                val clientDataJSON = credential.response.clientDataJSON
                val bytes = js("new Uint8Array(clientDataJSON)")
                var clientBinary = ""
                for (i in 0 until bytes.length) {
                    clientBinary += js("String.fromCharCode(bytes[i])")
                }
                val clientBase64 = toBase64Url(window.btoa(clientBinary))
                val jsonStr = """{"id":"$idBase64","type":"public-key","rawId":"$rawIdBase64","response":{"clientDataJSON":"$clientBase64"}}"""
                val body = "credential=${encodeURIComponent(jsonStr)}"
                window.asDynamic().fetch("/api/auth/authenticate", js("({method: 'POST', headers: {'Content-Type': 'application/x-www-form-urlencoded'}, body: body})")).then { res ->
                    res.unsafeCast<dynamic>().json().then<dynamic> { json -> json }
                }
            }
    }

    private fun parseAssertionOptions(options: dynamic): dynamic {
        options.challenge = base64ToArrayBuffer(options.challenge as String)
        val allowCredentials = options.allowCredentials
        if (allowCredentials != null && allowCredentials.length > 0) {
            for (i in 0 until allowCredentials.length) {
                allowCredentials[i].id = base64ToArrayBuffer(allowCredentials[i].id)
            }
        }
        return options
    }

    private fun getRegistrationOptions(email: String, displayName: String): Promise<dynamic> {
        val language = window.localStorage.getItem("preferredLanguage") ?: "en"
        val body = "email=${encodeURIComponent(email)}&displayName=${encodeURIComponent(displayName)}&language=${encodeURIComponent(language)}"
        console.log("Requesting registration options with: email=$email, displayName=$displayName, language=$language")
        return window.asDynamic().fetch("/api/auth/registration-options", js("({method: 'POST', headers: {'Content-Type': 'application/x-www-form-urlencoded'}, body: body})")).then { res ->
            console.log("Registration options response status: " + res.status)
            if (res.ok) {
                res.unsafeCast<dynamic>().json().then<dynamic> { json -> json }
            } else {
                res.json().then { json ->
                    val errorMsg = if (json != null && json.error != undefined) json.error as? String else "Registration failed"
                    console.log("Registration options error: " + errorMsg)
                    val error = js("new Error('')")
                    error.message = errorMsg
                    throw error
                }
            }
        }
    }

    private fun getAuthenticationOptions(): Promise<dynamic> {
        return apiFetch("/api/auth/assertion-options")
    }

    private fun parseCreationOptions(options: dynamic): dynamic {
        val pk = options.publicKey
        val userEntity = js("({})")
        userEntity.id = base64ToArrayBuffer(pk.user.id)
        userEntity.name = pk.user.name
        userEntity.displayName = pk.user.displayName
        val converted = js("({})")
        converted.rp = pk.rp
        converted.user = userEntity
        converted.challenge = base64ToArrayBuffer(pk.challenge)
        converted.pubKeyCredParams = pk.pubKeyCredParams
        val excludeList = pk.excludeCredentials
        if (excludeList != null) {
            val convertedList = js("[]")
            for (i in 0 until excludeList.length) {
                val exc = excludeList[i]
                val excConverted = js("({})")
                excConverted.id = base64ToArrayBuffer(exc.id)
                excConverted.type = exc.type
                convertedList.push(excConverted)
            }
            converted.excludeCredentials = convertedList
        }
        converted.timeout = pk.timeout
        converted.attestation = pk.attestation
        converted.extensions = pk.extensions
        return js("({publicKey: converted})")
    }

    private fun toBase64Url(base64: String): String {
        return base64.replace("+", "-").replace("/", "_").replace("=", "")
    }

    private fun base64ToArrayBuffer(base64: String): dynamic {
        val standardBase64 = base64.replace("-", "+").replace("_", "/")
        val padded = standardBase64 + "==".substring(0, (4 - standardBase64.length % 4) % 4)
        val binary = window.atob(padded)
        val len = binary.length
        val bytes = js("new Uint8Array(len)")
        for (i in 0 until len) {
            js("bytes[i] = binary.charCodeAt(i)")
        }
        return bytes.buffer
    }

    private fun arrayBufferToBase64(buffer: dynamic): String {
        val bytes = js("new Uint8Array(buffer)")
        var binary = ""
        for (i in 0 until bytes.length) {
            binary += js("String.fromCharCode(bytes[i])")
        }
        return toBase64Url(window.btoa(binary))
    }
}